### ISSUE TITLE
ICE when using extern enum variant as a type

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#31163]
The following and various modifications of it cause an ICE. It doesn't need to be in a pattern. It also crashes if it's a value. Normal error is reported if the enum is local.

```rust
extern crate rustc;

use rustc::lint::*;

fn main() {
}

enum X {
    A,
    Y(i32),
    B,
}
pub use self::X::*;

fn panics() {
    let Y::Uiae(asdf) = 99;
    let Node::Garblegab(Blubber) = 42;
}
```

backtrace (not of the simplified case above, but of the original case I had)

```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', ../src/libcore/option.rs:330
stack backtrace:
   1:     0x7f1ed23419d0 - sys::backtrace::tracing::imp::write::h188951552cfde9c3tnu
   2:     0x7f1ed23486c5 - panicking::default_handler::_<closure>::closure.42740
   3:     0x7f1ed2348253 - panicking::default_handler::h00ef84220a33c4b5VEy
   4:     0x7f1ed2311a16 - sys_common::unwind::begin_unwind_inner::h48700dfae980863d0ft
   5:     0x7f1ed23123d8 - sys_common::unwind::begin_unwind_fmt::h63b701b0b99330e96et
   6:     0x7f1ed233f5e1 - rust_begin_unwind
   7:     0x7f1ed238ec1f - panicking::panic_fmt::h48ff011da22960abnYL
   8:     0x7f1ed238f718 - panicking::panic::h240e4293bef76608UWL
   9:     0x7f1ed17171b8 - astconv::finish_resolving_def_to_ty::h8347503449ade6545aw
  10:     0x7f1ed16d4059 - check::resolve_ty_and_def_ufcs::h957cdf88801cdc79hSs
  11:     0x7f1ed171f6be - check::_match::check_pat_enum::hdf4b278a621d5ac2Csb
  12:     0x7f1ed16a6eb5 - check::_match::check_pat::hafcb0a62515bb61cUSa
  13:     0x7f1ed1720bd2 - check::_match::check_pat_enum::hdf4b278a621d5ac2Csb
  14:     0x7f1ed16a6eb5 - check::_match::check_pat::hafcb0a62515bb61cUSa
  15:     0x7f1ed16d95bf - check::_match::check_match::hff470f5bf3775f8dEib
  16:     0x7f1ed16f3db7 - check::check_expr_with_unifier::h13440647592506116803
  17:     0x7f1ed16a8f2e - check::check_block_with_expected::h7c3f25fc25a53651q4s
  18:     0x7f1ed16fd530 - check::check_expr_with_unifier::h5881387110198114312
  19:     0x7f1ed1721fbc - check::_match::check_match::_<closure>::closure.39088
  20:     0x7f1ed16d98d1 - check::_match::check_match::hff470f5bf3775f8dEib
  21:     0x7f1ed16f3db7 - check::check_expr_with_unifier::h13440647592506116803
  22:     0x7f1ed16a8f2e - check::check_block_with_expected::h7c3f25fc25a53651q4s
  23:     0x7f1ed16fd530 - check::check_expr_with_unifier::h5881387110198114312
  24:     0x7f1ed1721fbc - check::_match::check_match::_<closure>::closure.39088
  25:     0x7f1ed16d98d1 - check::_match::check_match::hff470f5bf3775f8dEib
  26:     0x7f1ed16f3db7 - check::check_expr_with_unifier::h13440647592506116803
  27:     0x7f1ed16a8f2e - check::check_block_with_expected::h7c3f25fc25a53651q4s
  28:     0x7f1ed169f70d - check::check_fn::h135d6c9abf3bab65HSo
  29:     0x7f1ed169c268 - check::check_bare_fn::hfe446a272ffadfd45Ho
  30:     0x7f1ed1699b64 - check::check_item_body::he654aa3e6f6fa1c1M7o
  31:     0x7f1ed1690628 - check::check_item_bodies::hce9e48e946ac1d25SFo
  32:     0x7f1ed168804e - check_crate::ha5b60b1f774b6def6KC
  33:     0x7f1ed286a52c - driver::phase_3_run_analysis_passes::_<closure>::closure.27263
  34:     0x7f1ed284d247 - middle::ty::context::ctxt<'tcx>::create_and_enter::h17072126803305727226
  35:     0x7f1ed28487a3 - driver::phase_3_run_analysis_passes::h7354846897191145749
  36:     0x7f1ed281c8b7 - driver::compile_input::h2880b830bdcda381uca
  37:     0x7f1ed280e95b - run_compiler::h6f82177125fbaba7exc
  38:     0x7f1ed280b751 - sys_common::unwind::try::try_fn::h11177386864859487898
  39:     0x7f1ed233f578 - __rust_try
  40:     0x7f1ed2336fdb - sys_common::unwind::try::inner_try::haf82437c0d63d3ffyct
  41:     0x7f1ed280bab0 - boxed::F.FnBox<A>::call_box::h3629726473297340986
  42:     0x7f1ed2346c43 - sys::thread::Thread::new::thread_start::h97374b34f63b05ddRBx
  43:     0x7f1ecae990a3 - start_thread
  44:     0x7f1ed1fbf04c - clone
  45:                0x0 - <unknown>
```

**Comment by mitaa** [comment]
This looks like #30535 and should be fixed in nightly.
(I tested the current nightly locally and it seems to be fixed)

**Comment by arielb1** [comment]
fixed on master
